### PR TITLE
Attachment without tools.jar

### PR DESCRIPTION
**Pull request by felixbarny**
This PR leverages the recently introduced emulated attach capabilities of byte buddy. This allows attachment on Windows, Solaris and Posix for HotSpot and Open J9 VMs, without relying on the `tools.jar`, `jdk.attach` module.

This also implements scanning for running JVMs without needing the `jps` command. That, however, only works for HotSpot VMs.

All in all, this makes the attacher usable in more environments like, for example, minimal docker containers which don't include the `tools.jar`, the `jdk.attach` module or `jps`.

Not sure yet how to package that.
My first idea was to not include any dependencies (byte-buddy, jna) and make them being resolved transitively when using the attachment via the [API](https://www.elastic.co/guide/en/apm/agent/java/current/setup-attach-api.html). Additionally, we'd distribute a jar containing all dependencies for standalone [`java -jar`](https://www.elastic.co/guide/en/apm/agent/java/current/setup-attach-cli.html) usage.

Some gotchas:
Spring Boot comes with a [parent pom](https://repo1.maven.org/maven2/org/springframework/boot/spring-boot-dependencies/2.1.6.RELEASE/spring-boot-dependencies-2.1.6.RELEASE.pom) which pins the versions of byte-buddy to 1.9.12 and JNA to 4.5.2. These versions don't allow for the emulated attach. The problem is that JNA can't be shaded to overcome this problem. See also https://github.com/java-native-access/jna/issues/679. If JNA is on the classpath, but with an older version, that leads this exception:
 ```
Caused by: java.lang.NoSuchMethodError: com.sun.jna.Native.load(Ljava/lang/String;Ljava/lang/Class;)Lcom/sun/jna/Library;
	at net.bytebuddy.agent.VirtualMachine$ForHotSpot$Connection$ForJnaPosixSocket$Factory.<init>(VirtualMachine.java:748)
```

Another idea would be to always include and shade byte buddy and to transitively resolve jna. We should add a check so that the emulated attach is only tried when JNA is available in the correct version. That check could actually be in byte buddy itself I think (https://github.com/raphw/byte-buddy/pull/704). 

relates to #638, elastic/apm-agent-java#647

### Checklist

- [x] Implement code
- [x] Add tests
- [x] Update documentation
- [ ] Update [CHANGELOG.md](CHANGELOG.md)
- [ ] Update [supported-technologies.asciidoc](docs/supported-technologies.asciidoc)
- [ ] Added an API method or config option? Document in which version this will be introduced.

### CODE REVIEW

**Comment by felixbarny** [comment]
Byte Buddy 1.10.1 will be JNA 4 compatible. So we should find out a way to shade Byte Buddy (to make sure we are using the exact required version, with the features and apis we depend on) itself but use JNA as a normal, transitively resolved dependency.

**Comment by felixbarny** [comment]
Seems like runtime attachment in combination with OSGi containers (Like JBoss, WildFly, etc.) is not possible at all at the moment :disappointed:
The reason seems to be that we are setting the bootdelegation system property too late - after the OSGi classloaders have been initialized.

**Comment by felixbarny** [comment]
Jenkins run the tests please

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java [hunk 1]
```diff
@@ -93,6 +95,12 @@ public static void onEnterServletService(@Advice.Argument(0) ServletRequest serv
             servletRequest.getDispatcherType() == DispatcherType.REQUEST &&
             !Boolean.TRUE.equals(excluded.get())) {
 
+            ServletContext servletContext = servletRequest.getServletContext();
+            if (servletContext != null) {
+                // this makes sure service name discovery also works when attaching at runtime
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java]: Why didn't we do it on the first place? We can lose metrics events for this service on startup or something?

**Comment by eyalkoren** [comment]
Jenkins run the tests

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/AnnotationValueOffsetMappingFactory.java [hunk 2]
```diff
@@ -53,7 +53,7 @@
         return new Advice.OffsetMapping() {
             @Override
             public Target resolve(TypeDescription instrumentedType, MethodDescription instrumentedMethod, Assigner assigner, Advice.ArgumentHandler argumentHandler, Sort sort) {
-                return Target.ForStackManipulation.of(getAnnotationValue(instrumentedMethod, annotation.loadSilent()));
+                return Target.ForStackManipulation.of(getAnnotationValue(instrumentedMethod, annotation.load()));
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/AnnotationValueOffsetMappingFactory.java]: `loadSilent()` was deprecated has been removed from Byte Buddy. It has the same semantics

File: apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ApmServerConfigurationSource.java [hunk 3]
```diff
@@ -61,7 +61,7 @@
     private static final int DEFAULT_POLL_DELAY_SEC = (int) TimeUnit.MINUTES.toSeconds(5);
     private static final Pattern MAX_AGE = Pattern.compile("max-age\\s*=\\s*(\\d+)");
     private final Logger logger;
-    private final DslJson<Object> dslJson = new DslJson<>();
+    private final DslJson<Object> dslJson = new DslJson<>(new DslJson.Settings<>());
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ApmServerConfigurationSource.java]: It's in master already

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java (hunk 1 above)
Comment (felixbarny) [reply apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java]: I don't think we can loose events with that approach. Only doing name discovery on Servlet#init had the benefit that we didn't have to do it on each transaction. But the filtering based on looking up in a set if the name has already been determined for a given servlet path should be quick enough.

**Comment by felixbarny** [comment]
closing in favor of #791

**Pull request closed without merging**